qa/setup_packages.py

```diff
@@ -16,34 +16,433 @@
+
+            Parameters
+            ----------
+            `cuda_version`: str
+                Cuda version used for this query
+        """
+        cuda_version = self.max_cuda_version(cuda_version)
+        ret = []
+        for v in self.versions[cuda_version]:
+            vers = self.get_pyvers_name(v, cuda_version)
+            if vers != "":
+                ret.append(vers)
+        return ret
+
+    def get_install_string(self, idx, cuda_version=None):
+        """Obtains installation string that pip should accept for version at
+            a given index with a given cuda_version
+
+            It checks what package is available and is compatible with the available platform
+
+            Parameters
+            ----------
+            `idx`: int
+                Index of version to retrive. If index is beyond 0-num_ov_versions-1 range
+                it is clamped to it
+            `cuda_version`: str
+                Cuda version used for this query
+        """
+        return "{version}".format(version=self.get_version(idx, cuda_version))
+
+    def test_request(self, url):
+        """Checks if a provided url is available
+
+            Parameters
+            ----------
+            `url`: str
+                Package url to be tested.
+        """
+        url = url.split("://")
+        url[-1] = urllib.parse.quote(url[-1])
+        url = "://".join(url)
+        request = Request(url)
+        request.get_method = lambda : 'HEAD'
+        try:
+            _ = urlopen(request)
+            return url
+        except HTTPError:
+            return None
+
+    def get_pyvers_name(self, url, cuda_version):
+        """Checks if a provided url is available for a given cuda version
+
+            It checks what package is available and is compatible with the available platforms
+            returned by the pip
+
+            Parameters
+            ----------
+            `url`: str
+                Package url to be tested. `{cuda_v}` is replaced by cuda_version and  `{platform}`
+                by the platform tag
+            `cuda_version`: str
+                Cuda version used for this query
+        """
+        if isinstance(p.get_supported()[0], tuple):
+            # old PIP returns tuple
+            for py_ver in [(x, y, z) for (x, y, z) in p.get_supported() if y != 'none' and 'any' not in y]:
+                py_ver = "-".join(py_ver)
+                ret = self.test_request(url.format(platform=py_ver, cuda_v=cuda_version))
+                if ret:
+                    return ret
+        else:
+            # new PIP returns object
+            for py_ver in [tag for tag in p.get_supported() if tag.abi != 'none' and tag.platform != 'any']:
+                py_ver = str(py_ver)
+                ret = self.test_request(url.format(platform=py_ver, cuda_v=cuda_version))
+                if ret:
+                    return ret
+        return ""
+
+all_packages = [PlainPackage("opencv-python", ["4.2.0.32"]),
```

Comment:
Everything below seems like a nice improvement.

Reply:
?

Reply:
Just wanted to say, that it's more readable than the previous one.

Reply:
Oh, ok.